lib/fluent/plugin/in_tail.rb

```diff
@@ -494,9 +500,9 @@ def on_notify
             begin
               while true
                 if @buffer.empty?
-                  @io.read_nonblock(2048, @buffer)
+                  @io.readpartial(2048, @buffer)
```

Comment:
Almost yes, and in this case the difference is not an issue (or better than original).

#### read_nonblock(nbytes) behaves:
* If Ruby's internal IO buffer has chars, read (0, nbytes] bytes from it, and return it.
* Otherwise call read(2) **with** setting the io as O_NONBLOCK.
  * If it can read string, return it.
  * Otherwise, raise an exception Errno::EAGAIN or Errno::EWOULDBLOCK

#### read_partial behaves
* If Ruby's internal IO buffer has chars, read (0, nbytes] bytes from it, and return it.
* Otherwise call read(2) **without** setting the io as O_NONBLOCK.
  * If it can read string, return it even if the length is less than nbytes.
  * Otherwise, wait until it gets at least one byte char or EOF.

#### read(nbytes)
* If Ruby's internal IO buffer has at least 2048 chars, read and return it.
* Otherwise read rest of bytes with read(2) with blocking.

### In this case
The code is in infinite loop with reading `@io`. Therefore using read_nonblock without something blocking call like select may consume extra CPU. IO#readpartial realize such efficient logic without struggling IO.select. (Moreover read_nonblock may raise EWOULDBLOCK if there's enough bytes. I don't know why previous code works well. (Won't happen such case in current use case?))